api-flask/app/zonal_stats.py

```diff
@@ -62,19 +113,80 @@ def _group_zones(zones, group_by):
     return output_file
 
 
+def _create_geoms(geojson_dict):
+    """Read and parse geojson dictionary geometries into shapely objects."""
+    geometries = {}
+    for feature in geojson_dict.get('data').get('features'):
+        key = feature.get('properties').get(geojson_dict.get('key'))
+        geom = shape(feature.get('geometry'))
+
+        geometries[key] = geom
+
+    return geometries
+
+
+def _compute_wfs_stats(zones_dict, wfs_response, geotiff, stats_params):
+    """Compute stats for each individual polygon and geotiff image."""
+    wfs_geoms = [shape(f.get('geometry')) for f in wfs_response.get('features')
+                 if f.get('geometry').get('type') in ['MultiPolygon', 'Polygon']]
+    zones_geoms = _create_geoms(zones_dict)
+
+    results = []
+    keys = []
+    for key, value in zones_geoms.items():
+        filtered = [value.intersection(g) for g in wfs_geoms if value.intersects(g)]
+
+        if len(filtered) == 0:
+            continue
+        results.append(cascaded_union(filtered))
+        keys.append({zones_dict['key']: key})
+
+    stats = zonal_stats(results, geotiff, **stats_params)
+    parsed_stats = []
+    for stat in stats:
+        props = {k: v if v is None else int(v) for k, v in stat['properties'].items()}
+        props['geometry'] = stat['geometry']
+        parsed_stats.append(props)
+
+    output = [{**key, **stat}
+              for stat, key in zip(parsed_stats, keys)]
+
+    return output
+
+
 @timed
 def calculate_stats(
     zones,
     geotiff,
     group_by=None,
     stats=DEFAULT_STATS,
     prefix='stats_',
-    geojson_out=False
+    geojson_out=False,
+    wfs_response=None
 ):
     """Calculate stats."""
     if group_by:
         zones = _group_zones(zones, group_by)
 
+    if wfs_response:
+        zones_key = group_by if group_by else 'TS'
```

Comment:
'TS' is specific to Myanmar. I'm not sure of a use case where the group_by parameter would not be explicitly included in the call to the zonal stats analysis. We should look into making this a mandatory parameter

Reply:
I just modified the code in https://github.com/WFP-VAM/prism-frontend/pull/186/commits/7e2378f868750bfe03484e0a364181edcb35cd18 which removes the group_by check and simplifies code drastically. cc @ericboucher